### PR TITLE
Add ajax

### PR DESCRIPTION
**Pull request by rineshabana**
In this branch we practised how we can use ajax to make our application behave like single page application.
To do this we made use of the `remote: true` on the forms and the links this sends out a `xhr` request which is handled through the javascript views `*.js.erb`.
We also used `noty` for viewing various notification message.

### CODE REVIEW

File: app/views/user_address/index.html.erb [hunk 1]
```diff
@@ -18,7 +16,4 @@
                 </div>
             </div>
             <br>
-        </div>
-    
-    <% end %>
-</div>
\ No newline at end of file
+    <% end %>
```
Comment (stingrayzboy) [inline app/views/user_address/index.html.erb]: Please add a new line at the end of the code.

File: config/webpack/environment.js [hunk 2]
```diff
@@ -1,3 +1,10 @@
 const { environment } = require('@rails/webpacker')
 
 module.exports = environment
+const webpack = require('webpack')
+environment.plugins.prepend('Provide',
+  new webpack.ProvidePlugin({
+    $: 'jquery/src/jquery',
+    jQuery: 'jquery/src/jquery'
+  })
+)
```
Comment (stingrayzboy) [inline config/webpack/environment.js]: Please add new line here

File: app/controllers/user_address_controller.rb [hunk 3]
```diff
@@ -16,19 +16,25 @@ def edit; end
   def create
     @address = UserAddress.new(address_params)
     @address.user_profile = current_user.user_profile
-    raise unless @address.save
-
-    redirect_to user_address_index_path
+    if @address.save
+      flash[:success] = t('address.create.success')
```
Comment (stingrayzboy) [inline app/controllers/user_address_controller.rb:20]: be consistent with the flash types if using success/ failures dont use notice.

File: app/views/user_address/_address-card.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        <p class="card-text">Address Line 2 : <%= addr.address_line2%></p>
+        <p class="card-text">Address Line 3 : <%= addr.address_line3%></p>
+        <p class="card-text">City : <%= addr.city%></p>
+        <p class="card-text">State : <%= addr.state%></p>
+        <p class="card-text">Pin Code : <%= addr.pincode%></p>
+        <p class="card-text">Contact Number : <%= show_contact_no(addr)%></p>
+        <%= link_to "Edit", edit_user_address_path(id: addr.id), class: "btn btn-warning" ,remote:true%>
+        <%= link_to "Delete", destroy_user_address_path(id: addr.id), class: "btn btn-danger", method: :delete, remote: true %>
+    </div>
+</div>
```
Comment (stingrayzboy) [inline app/views/user_address/_address-card.html.erb]: new line here.

File: app/views/user_address/_index.html.erb [hunk 5]
```diff
@@ -1,19 +1,6 @@
...
-                    <%= link_to "Delete", destroy_user_address_path(id: address.id), class: "btn btn-danger", method: :delete %>
-                </div>
-            </div>
-            <br>
-    <% end %>
\ No newline at end of file
+<% addresses.each do |address| %>
+        <%= render partial: 'address-card', locals:{addr:address}%>
+        <br>
+<% end %>
```
Comment (stingrayzboy) [inline app/views/user_address/_index.html.erb]: new line ??

File: app/views/user_address/edit.js.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+$("#formport").html("<%= j render partial: 'edit'%>");
```
Comment (stingrayzboy) [inline app/views/user_address/edit.js.erb:1]: new line

File: app/views/user_address/new.js.erb [hunk 7]
```diff
@@ -0,0 +1 @@
+$("#formport").html("<%= j render partial: 'new'%>");
```
Comment (stingrayzboy) [inline app/views/user_address/new.js.erb]: new line.

File: app/controllers/user_address_controller.rb [hunk 8]
```diff
@@ -4,7 +4,7 @@ class UserAddressController < ApplicationController
 
   def index
     @addresses = current_user.user_addresses
-    flash[:notice] = t('create_profile') if current_user.user_profile.nil?
+    flash[:notice] = t('profile.create.message') if current_user.user_profile.nil?
```
Comment (stingrayzboy) [inline app/controllers/user_address_controller.rb:7]: notice is ok to be used here.

File: app/views/user_profile/_form.html.erb [hunk 9]
```diff
@@ -1,43 +1,43 @@
...
+            <div class="mb-3">
+                <%= f.label :contact_number %>
+                <%= f.text_field :contact_number,class: "form-control" %>
             </div>
-        <% end %>
\ No newline at end of file
+                <%= f.submit "Sumbit", class: "btn btn-primary" %>
+        </div>
+    </div>
+<% end %>
```
Comment (stingrayzboy) [inline app/views/user_profile/_form.html.erb:44]: new line.

File: app/views/user_profile/create.js.erb [hunk 10]
```diff
@@ -0,0 +1,3 @@
+new Noty({theme:'metroui', timeout: 1000, type: 'success',layout: 'topCenter',text: '<%= flash[:notice]%>'}).show();
+$('#formport').html("");
+$('#viewport').html("<%= j render partial:'show'%>")
```
Comment (stingrayzboy) [inline app/views/user_profile/create.js.erb]: new line

File: app/views/user_profile/edit.js.erb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+$('#viewport').html("");
+$('#formport').html("<%= j render partial: 'form', locals: { obj: @profile, routing: {controller: :user_profile, action: :update} }%>")
```
Comment (stingrayzboy) [inline app/views/user_profile/edit.js.erb]: new line

File: app/views/user_profile/show.js.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<%if @create == true%>
+    new Noty({theme:'metroui', timeout: 1000, type: 'error',layout: 'topCenter',text: '<%= flash[:notice]%>'}).show();
+    $('#viewport').html("");
+    $('#formport').html("<%= j render partial: 'form', locals: { obj: @profile, routing: {controller: :user_profile, action: :create} }%>")
+<%else%>
+    $('#formport').html("");
+    $('#viewport').html("<%= j render partial:'show'%>")
+<%end%>
```
Comment (stingrayzboy) [inline app/views/user_profile/show.js.erb]: new line.

File: app/views/user_address/_form.html.erb [hunk 13]
```diff
@@ -42,6 +42,7 @@
                     <%= f.text_field :contact_number,class: "form-control" %>
                 </div>
                     <%= f.submit "Sumbit", class: "btn btn-primary" %>
+                    <button class="btn btn-danger" onClick='closeForm()'>Cancel</button>
             </div>
         </div>
     <% end %>
```
Comment (stingrayzboy) [inline app/views/user_address/_form.html.erb]: new line

**Review by stingrayzboy — CHANGES_REQUESTED** [review]
Add the new lines

File: app/views/user_address/index.html.erb (hunk 1 above)
Comment (rineshabana) [reply app/views/user_address/index.html.erb]: ok

File: config/webpack/environment.js (hunk 2 above)
Comment (rineshabana) [reply config/webpack/environment.js]: ok

File: app/controllers/user_address_controller.rb (hunk 3 above)
Comment (rineshabana) [reply app/controllers/user_address_controller.rb:20]: ok

File: app/views/user_address/_index.html.erb (hunk 5 above)
Comment (rineshabana) [reply app/views/user_address/_index.html.erb]: ok

File: app/views/user_profile/create.js.erb (hunk 10 above)
Comment (rineshabana) [reply app/views/user_profile/create.js.erb]: ok

File: app/views/user_profile/show.js.erb (hunk 12 above)
Comment (rineshabana) [reply app/views/user_profile/show.js.erb]: ok

File: app/views/user_address/_address-card.html.erb (hunk 4 above)
Comment (rineshabana) [reply app/views/user_address/_address-card.html.erb]: ok

File: app/views/user_profile/_form.html.erb (hunk 9 above)
Comment (rineshabana) [reply app/views/user_profile/_form.html.erb:44]: ok

File: app/views/user_profile/edit.js.erb (hunk 11 above)
Comment (rineshabana) [reply app/views/user_profile/edit.js.erb]: ok

File: app/views/user_address/_form.html.erb (hunk 13 above)
Comment (rineshabana) [reply app/views/user_address/_form.html.erb]: ok

File: app/views/user_address/edit.js.erb (hunk 6 above)
Comment (rineshabana) [reply app/views/user_address/edit.js.erb:1]: ok

File: app/views/user_address/new.js.erb (hunk 7 above)
Comment (rineshabana) [reply app/views/user_address/new.js.erb]: ok

File: app/controllers/user_address_controller.rb (hunk 8 above)
Comment (rineshabana) [reply app/controllers/user_address_controller.rb:7]: ok

**Review by stingrayzboy — CHANGES_REQUESTED** [review]
Change the commit message from Minor Refactoring instead of Minot Refactoring

**Review by stingrayzboy — APPROVED** [review]
Looks Good.
Well Done! 👏

**Pull request merged**